src/Foundatio/Caching/InMemoryCacheClient.cs

```diff
@@ -189,26 +195,33 @@ public Task<int> RemoveByPrefixAsync(string prefix)
 
     internal void RemoveExpiredKey(string key, bool sendNotification = true)
     {
-        _logger.LogDebug("Removing expired cache entry {Key}", key);
+        // Consideration: We could reduce the amount of calls to this by updating ExpiresAt and only having maintenance remove keys.
+        if (_memory.TryGetValue(key, out var existingEntry) && existingEntry.ExpiresAt < SystemClock.UtcNow)
+        {
+            if (_memory.TryRemove(key, out var removedEntry))
+            {
+                if (removedEntry.ExpiresAt >= SystemClock.UtcNow)
+                    throw new Exception("Removed item was not expired");
 
-        if (_memory.TryRemove(key, out _))
-            OnItemExpired(key, sendNotification);
+                _logger.LogDebug("Removing expired cache entry {Key}", key);
+                OnItemExpired(key, sendNotification);
+            }
+        }
     }
 
     public Task<CacheValue<T>> GetAsync<T>(string key)
     {
         if (String.IsNullOrEmpty(key))
             throw new ArgumentNullException(nameof(key), "Key cannot be null or empty");
 
-        if (!_memory.TryGetValue(key, out var cacheEntry))
+        if (!_memory.TryGetValue(key, out var existingEntry))
         {
             Interlocked.Increment(ref _misses);
             return Task.FromResult(CacheValue<T>.NoValue);
         }
 
-        if (cacheEntry.ExpiresAt < SystemClock.UtcNow)
+        if (existingEntry.ExpiresAt < SystemClock.UtcNow)
         {
-            RemoveExpiredKey(key);
```

Comment:
Just let maintenance remove the item to cut down on the number concurrent updates.